sense/finetuning.py

```diff
@@ -268,21 +270,41 @@ def extract_features(path_in, net, num_layers_finetune, use_gpu, num_timesteps=1
 
 
 def training_loops(net, train_loader, valid_loader, use_gpu, num_epochs, lr_schedule, label_names, path_out,
-                   temporal_annotation_training=False):
+                   temporal_annotation_training=False, save_checkpoints=False, frequency=5,
+                   load_checkpoint=False, resume_epoch=-1, change_layers=None, checkpoint_classifier=None):
     criterion = nn.CrossEntropyLoss()
     optimizer = optim.Adam(net.parameters(), lr=0.0001)
 
+    # load custom optimizer
+    if isinstance(load_checkpoint, str):
+        if change_layers == "No change" or change_layers == "N":
+            optimizer.load_state_dict(checkpoint_classifier['optimizer'])
+        # remove all checkpoints and epoch restart at 0 if training from a intermediate checkpoint
+        if load_checkpoint != '':
+            cdir = os.path.dirname(load_checkpoint)
+            filelist = [f for f in os.listdir(cdir)]
+            for f in filelist:
+                os.remove(os.path.join(cdir, f))
+
     best_state_dict = None
     best_top1 = 0.
     best_loss = 9999
 
-    for epoch in range(num_epochs):  # loop over the dataset multiple times
+    # if resume training from the last checkpoint, epoch will continue at epoch + 1
+    # otherwise, it will start at 0
+    for epoch in range(resume_epoch + 1, num_epochs + resume_epoch + 1):  # loop over the dataset multiple times
         new_lr = lr_schedule.get(epoch)
         if new_lr:
             print(f"update lr to {new_lr}")
             for param_group in optimizer.param_groups:
                 param_group['lr'] = new_lr
 
+        # rename the last saved checkpoint
+        if epoch % frequency == 1 and \
+                os.path.exists(os.path.join(path_out, "checkpoints/", "classifier%g.checkpoint" % (epoch - 1))):
```

Comment:
can we use a f-string here ? would make it easier to read

Reply:
> can we use a f-string here ? would make it easier to read

yes, I'll try use f-string format here, thanks